lib/vector-core/src/event/log_event.rs

```diff
@@ -19,38 +23,104 @@ use super::{
 };
 use crate::{config::log_schema, event::MaybeAsLogMut, ByteSizeOf};
 
-#[derive(Clone, Debug, PartialEq, PartialOrd, Deserialize)]
-pub struct LogEvent {
+#[derive(Debug, Deserialize)]
+struct Inner {
     #[serde(flatten)]
-    fields: Arc<Value>,
+    fields: Value,
 
     #[serde(skip)]
-    metadata: EventMetadata,
+    size_cache: Atomic<Option<NonZeroUsize>>,
 }
 
-impl LogEvent {
-    pub fn metadata(&self) -> &EventMetadata {
-        &self.metadata
+impl Inner {
+    fn invalidate(&self) {
+        self.size_cache.store(None, Ordering::Relaxed);
     }
+}
 
-    pub fn metadata_mut(&mut self) -> &mut EventMetadata {
-        &mut self.metadata
+impl ByteSizeOf for Inner {
+    fn size_of(&self) -> usize {
+        self.size_cache
+            .load(Ordering::Relaxed)
+            .unwrap_or_else(|| {
+                let size = size_of::<Self>() + self.allocated_bytes();
+                let size = NonZeroUsize::new(size).expect("Size cannot be zero");
```

Comment:
This is probably true right now, but I could see, in the future, having an "empty" event (that is, one with no fields) have a size of 0.

Reply:
That is, I might relax this constraint.

Reply:
The thing is, `Self` will always have size > 0, so that plus the allocated bytes will always be non-zero. Overflow is also not possible, because `usize` has a range the same as the address space.

I had a comment to this effect in an earlier version, don't know why I dropped it.